inc/functions.php

```diff
@@ -569,49 +627,44 @@ function my_mail($to, $subject, $message, $from="", $charset="", $headers="", $k
 	global $mybb, $plugins;
 	static $mail;
 
-	// Does our object not exist? Create it
-	if(!is_object($mail))
-	{
-		require_once MYBB_ROOT."inc/class_mailhandler.php";
+	// Get our mail handler.
+	$mail = &get_mybb_mailhandler();
 
-		$plugins->run_hooks('my_mail_mailhandler_init', $mail);
-
-		if(!is_object($mail) || !($mail instanceof MailHandler))
+	$my_mail_parameters = compact("to", "subject", "message", "from", "charset", "headers", "keep_alive", "format", "message_text", "return_email");
+	$my_mail_parameters = $plugins->run_hooks('my_mail_parameters', $my_mail_parameters);
+	if(is_array($my_mail_parameters))
+	{
+		foreach($my_mail_parameters as $key => $value)
 		{
-			if($mybb->settings['mail_handler'] == 'smtp')
+			// Using extract() here may not be safe, so we work it out manually.
+			if(in_array($key, array("to", "subject", "message", "from", "charset", "headers", "keep_alive", "format", "message_text", "return_email")))
 			{
-				require_once MYBB_ROOT."inc/mailhandlers/smtp.php";
-				$mail = new SmtpMail();
-			}
-			else
-			{
-				require_once MYBB_ROOT."inc/mailhandlers/php.php";
-				$mail = new PhpMail();
+				$$key = $value;
 			}
 		}
 	}
+	$my_mail_parameters = compact("to", "subject", "message", "from", "charset", "headers", "keep_alive", "format", "message_text", "return_email");
 
-	// Using SMTP based mail
-	if($mybb->settings['mail_handler'] == 'smtp')
+	// If MyBB's built-in SMTP mail handler is used, set the keep alive bit accordingly.
+	if(isset($mybb->settings['mail_handler']) && $mybb->settings['mail_handler'] == 'smtp' && $keep_alive == true && !empty($mail->keep_alive))
 	{
-		if($keep_alive == true)
+		require_once MYBB_ROOT . "inc/class_mailhandler.php";
+		require_once MYBB_ROOT . "inc/mailhandlers/smtp.php";
+		if($mail instanceof MailHandler && $mail instanceof SmtpMail)
 		{
 			$mail->keep_alive = true;
 		}
 	}
 
-	// Using PHP based mail()
-	else
-	{
-		if($mybb->settings['mail_parameters'] != '')
-		{
-			$mail->additional_parameters = $mybb->settings['mail_parameters'];
-		}
-	}
-
-	// Build and send
+	// Build the mail message.
 	$mail->build_message($to, $subject, $message, $from, $charset, $headers, $format, $message_text, $return_email);
-	return $mail->send();
+
+	// Send the mail via this hook, if any.
+	// The hook would return true/false if the mail is sent or not, otherwise anything else that is none boolean.
+	$sent = $plugins->run_hooks('my_mail_send', $my_mail_parameters);
```

Comment:
The position of this hook could be adjusted, maybe before the `$mail->build_message`?